libgalois/src/DynamicBitset.cpp

```diff
@@ -96,90 +141,140 @@ katana::DynamicBitset::count() const {
+          size_t{0}, underlying_bitvec.size(), tid, nthreads);
+
       Integer index = cur_size;
       if (tid != 0) {
-        index += tPrefixBitCounts[tid - 1];
+        index += thread_prefix_bit_counts[tid - 1];
       }
 
-      for (Integer i = start; i < end; ++i) {
-        if (bitset.test(i)) {
-          offsets->at(index) = i;
-          ++index;
+      for (uint64_t bitvec_index = start; bitvec_index < end; ++bitvec_index) {
+        // get set bits and add
+        uint64_t current_num = underlying_bitvec[bitvec_index];
+        uint64_t offset =
+            bitvec_index * katana::DynamicBitset::kNumBitsInUint64;
+
+        if (current_num == 0) {
+          // nothing to add
+          continue;
+        }
+
+        if (current_num == 1) {
+          // trailing 0s indicate location of set bit
+          set_elements->at(index++) = offset + CountTrailingZeroes(current_num);
+        } else if (current_num == std::numeric_limits<uint64_t>::max()) {
+          // add all
+          for (size_t i = 0; i < katana::DynamicBitset::kNumBitsInUint64; i++) {
+            set_elements->at(index++) = offset + i;
+          }
+        } else {
+          // add only set parts
+          for (size_t i = 0; i < katana::DynamicBitset::kNumBitsInUint64; i++) {
+            if (bitset.test(offset + i)) {
+              set_elements->at(index++) = offset + i;
+            }
+          }
         }
       }
     });
   }
 }
-}  //namespace
+
+template <typename Integer>
+void
+ComputeOffsetsSerial(
+    const katana::DynamicBitset& bitset, std::vector<Integer>* set_elements) {
+  const katana::PODVector<katana::DynamicBitset::TItem>& underlying_bitvec =
+      bitset.get_vec();
+  set_elements->reserve(bitset.SerialCount());
+
+  // loop through each int of the bitset invidiually
+  for (size_t bit_index = 0; bit_index < underlying_bitvec.size();
+       bit_index++) {
+    uint64_t int_to_examine = underlying_bitvec[bit_index];
+
+    if (int_to_examine == 0) {
+      // nothing set, skip
+      continue;
+    }
+
+    uint64_t offset = bit_index * katana::DynamicBitset::kNumBitsInUint64;
+
+    // TODO(l-hoang) optimize 63 set bits case? (find only 0, add the rest)
+    // optimize special corner cases case
+    if (int_to_examine == 1) {
+      // find trailing zeros gets you the set bit without needing to loop over
+      // all 64 bits
+      set_elements->emplace_back(offset + CountTrailingZeroes(int_to_examine));
+    } else if (int_to_examine == std::numeric_limits<uint64_t>::max()) {
+      // all set: add all, no testing required
+      for (size_t i = 0; i < katana::DynamicBitset::kNumBitsInUint64; i++) {
+        set_elements->emplace_back(offset + i);
+      }
+    } else {
+      // loop through all; hope is that this is rare
+      // TODO(l-hoang) is there a better way to do this, i.e. is test required?
+      for (size_t i = 0; i < katana::DynamicBitset::kNumBitsInUint64; i++) {
+        if (bitset.test(offset + i)) {
```

Comment:
While we are splitting optimization hairs, I don't think it makes sense to call `bitset.test()`. You already have the relevant element of `bitvec_`, why not just compare your index against that instead of having the bitset redo the math you already did? You can look at `DynamicBitsetIterator` for an example of this. I have the same comment for the parallel version.